### DISCUSSION TITLE
perplexed about  /stevenarella/blocks/src/lib.rs

### DISCUSSION THREAD

**Discussion by Kezii** [iceiix/stevenarella#453] [General]
Is that a compiled array with all the 1.12 blocks? Is that the reason the world is on fire when joining an 1.16 server? How can we do better?

**Comment by iceiix** [comment]
@Kezii Yeah there is a [`define_blocks!`](https://github.com/iceiix/stevenarella/blob/master/blocks/src/lib.rs#L509) macro in stevens_blocks defining all the supported blocks, but [src/resources.rs](https://github.com/iceiix/stevenarella/blob/master/src/resources.rs#L29-L33) currently uses 1.12.2 assets. 

We could (and should) update the assets, I've started it here: https://github.com/iceiix/stevenarella/pull/71 but the problem is the 1.12->1.13 transition incompatibly changes several asset names. So this raises the question, should we commit to supporting only 1.13+ compatible resource packs, or also 1.7-1.12 resource packs?

This isn't a real problem yet but would be once we started supporting custom resource packs, https://github.com/iceiix/stevenarella/issues/399. I'd prefer supporting both formats, so a (say) 1.7.10 user could use 1.7.10 resource packs. Looking at https://minecraft.gamepedia.com/Resource_Pack, there are even more "resource pack versions". 1.8.8 is 1, 1.9 is 2 ("due to changes in the model system, such as item tags, multipart, and changes to display tags", didn't affect us), 1.11 is 3 ("due to the change that all files should have lowercase letters", didn't matter), and 1.13 (major change) is 4. 1.15 changed to 5 ("texture mechanic changes"), 1.16.2 to 6 ("changes to wall blocks"), and upcoming 1.17 will be 7. Haven't looked into these changes in depth, but they may require other changes on our end, which could introduce more resource pack version incompatibilities.

As for the fire blocks themselves, it is curious how these are shown instead of Block::Missing, I'm not sure what is going on there but it may be a bug in our update for [The Flattening](https://minecraft.gamepedia.com/Java_Edition_1.13/Flattening). Pre-flattening protocol versions, blocks are identified by their index in `define_blocks!` and the `data` short (from `get_hierarchical_data()`), post-flattening they are instead identified by `offset` (through `get_flat_offset()`. `by_vanilla_id` handles both cases, depending on protocol_version.

**Comment by iceiix** [comment]
@Kezii I investigated this problem some more, created an issue: https://github.com/iceiix/stevenarella/issues/467

The bigger problem besides outdated assets is that block states are not stable across versions.

From 1.13.2 to 1.14(.4?), new states were added so that "grass" shifted from block state [1041](https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.13/blocks.json#L2941) to [1341](https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.14.4/blocks.json#L2971), but this is not accounted for in blocks/src/lib.rs which still recognizes it as 1041 (it was slightly off due to another bug, just fixed: #468), hence the fire.

steven_blocks already does calculate the block state IDs based on the possible values (ordered using the `offset` macro token). So I think a way to fix this would be to add version awareness to the block states, so they are only enabled on supported versions, add the missing states, and ensure they line up as expected.

**Reply by iceiix** [reply]
OK I implemented this change in https://github.com/iceiix/stevenarella/commit/8ca12a69cecb5935abf71ac766e752b94ea09b30, not all blocks are defined yet but 1.14.4 - 1.16.4 are at least no longer on fire:

<img width="966" alt="Screen Shot 2021-01-13 at 6 50 55 PM" src="https://user-images.githubusercontent.com/43691553/104543090-9c17c600-55d9-11eb-9fd2-ab62deb86345.png">
<img width="966" alt="Screen Shot 2021-01-13 at 7 05 10 PM" src="https://user-images.githubusercontent.com/43691553/104543099-a20da700-55d9-11eb-83f1-79c3c9482376.png">

Some additional context on how `define_blocks!` works:

It is a Rust [macro](https://doc.rust-lang.org/book/ch19-06-macros.html), defined with `macro_rules!`, used to generate instances and method implementations for `enum Block`.

Various macro tokens are matched, most are optional but `props` (properties) and `model` are required for each block.

The blocks are numbered in the order they are defined in the file, but optionally with additional block states, expanding a block definition to 1 or more blocks.

Before [1.13/Flattening](https://minecraft.gamepedia.com/Java_Edition_1.13/Flattening), that is, 1.7.10 - 1.12.2, block IDs are "hierarchical" (as I call it), structured into a "block ID" and "data" value. The `data` token is set to an expression returning an option with 0 to 15, corresponding to the block data values. This is an `Option<usize>` so if it is `None` instead of `Some()`, the block is skipped from these versions. The default for `data` is `Some(0)`. Accessed by `block.get_hierarchical_data()`  to populate `hier: Vec<Option<Block>>` in `VanillaIDMap`. There are 16 data values for all blocks in these versions, even if they are not all defined. Example definition:

```rust
    Stone {
        props {
            variant: StoneVariant = [
                StoneVariant::Normal,
                StoneVariant::Granite,
                StoneVariant::SmoothGranite,
                StoneVariant::Diorite,
                StoneVariant::SmoothDiorite,
                StoneVariant::Andesite,
                StoneVariant::SmoothAndesite
            ],
        },
        data Some(variant.data()),
        model { ("minecraft", variant.as_string() ) },
    }
```

After the flattening (1.13.2 to, currently, 1.16.4), the `offset` token is used instead, populating `flat: Vec<Option<Block>>` of `VanillaIDMap`. Just like `data` is it an `Option<usize>`, with `None` skipping the block, but `Some()` adding it. If omitted, it defaults to data (as in `Stone`). Example where it is useful, in this block `snowy` doesn't exist pre-flattening, it does post-flattening and comes before the `!snowy` block (two blocks are defined in sequence: `Grass { snowy: true }`, and `Grass { snowy: false }`):

```rust
    Grass {
        props {
            snowy: bool = [false, true],
        },
        data { if snowy { None } else { Some(0) } },
        offset { if snowy { Some(0) } else { Some(1) } },
        model { ("minecraft", "grass") },
        variant format!("snowy={}", snowy),
        tint TintType::Grass,
        update_state (world, pos) => Block::Grass{snowy: is_snowy(world, pos)},
    }
```

`offset` defines additional variants of a block, but it is not restricted to 16 values, and can be more or less. Only as many block states are generated as requested, meaning ID shifts can occur as new variants are added with new protocol versions, similar to how packet IDs shift with new protocol versions. 

To account for this, https://github.com/iceiix/stevenarella/commit/8ca12a69cecb5935abf71ac766e752b94ea09b30 added a new `offsets` token in the `define_blocks!` macro, an expression similar to `offset` but accepting a `protocol_version` parameter. It can also return None to not generate a block with that value, so this is how version-specific blocks are specified. Here's the note block, which ballooned from 500 states (1.13.2) to 800 states (1.14+):

```rust
    NoteBlock {
        props {
            instrument: NoteBlockInstrument = [
                NoteBlockInstrument::Harp,
                NoteBlockInstrument::BaseDrum,
                NoteBlockInstrument::Snare,
                NoteBlockInstrument::Hat,
                NoteBlockInstrument::Bass,
                NoteBlockInstrument::Flute,
                NoteBlockInstrument::Bell,
                NoteBlockInstrument::Guitar,
                NoteBlockInstrument::Chime,
                NoteBlockInstrument::Xylophone,
                NoteBlockInstrument::IronXylophone,
                NoteBlockInstrument::CowBell,
                NoteBlockInstrument::Didgeridoo,
                NoteBlockInstrument::Bit,
                NoteBlockInstrument::Banjo,
                NoteBlockInstrument::Pling
            ],
            note: u8 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24],
            powered: bool = [true, false],
        },
        data if instrument == NoteBlockInstrument::Harp && note == 0 && powered { Some(0) } else { None },
        offsets |protocol_version| (instrument.offsets(protocol_version)
            .map(|offset| offset * (25 * 2) + ((note as usize) << 1) + if powered { 0 } else { 1 })),
        model { ("minecraft", "noteblock") },
    }
```

The specifics of the version can be defined inline, in the expression, but I pulled it out into this method of `NoteBlockInstrument`, so the new instruments only exist in `protocol_version >= 477`:

```rust
    fn offsets(self, protocol_version: i32) -> Option<usize> {
        match self {
            NoteBlockInstrument::Harp => Some(0),
            NoteBlockInstrument::BaseDrum => Some(1),
            NoteBlockInstrument::Snare => Some(2),
            NoteBlockInstrument::Hat => Some(3),
            NoteBlockInstrument::Bass => Some(4),
            NoteBlockInstrument::Flute => Some(5),
            NoteBlockInstrument::Bell => Some(6),
            NoteBlockInstrument::Guitar => Some(7),
            NoteBlockInstrument::Chime => Some(8),
            NoteBlockInstrument::Xylophone => Some(9),
            _ => {
                if protocol_version >= 477 {
                    match self {
                        NoteBlockInstrument::IronXylophone => Some(10),
                        NoteBlockInstrument::CowBell => Some(11),
                        NoteBlockInstrument::Didgeridoo => Some(12),
                        NoteBlockInstrument::Bit => Some(13),
                        NoteBlockInstrument::Banjo => Some(14),
                        NoteBlockInstrument::Pling => Some(15),
                        _ => unreachable!(),
                    }
                } else {
                    None
                }
            }
        }
    }
```

There's still a lot of work to be done to add all the missing block states for newer versions, but these changes at least fix the most egregious error of grass = fire.
